cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+
+    // immediately error and reject any discardable commits, i.e. the
+    // the empty wcc
+    for commit in &to_upload {
+        if commit.is_discardable(tx.repo_mut())? {
+            return Err(user_error_with_hint(
+                format!(
+                    "Refusing to upload commit {} because it is an empty commit with no \
+                     description",
+                    short_commit_hash(commit.id())
+                ),
+                "Perhaps you squashed then ran upload? Maybe you meant to upload the parent \
+                 commit instead (eg. @-)",
+            ));
+        }
+    }
+
+    let mut old_to_new: IndexMap<CommitId, Commit> = IndexMap::new();
+    for commit_id in to_upload.iter().map(|c| c.id()).rev() {
+        let original_commit = store.get_commit(commit_id).unwrap();
+        let description = original_commit.description().to_owned();
+        let trailers = parse_description_trailers(&description);
+
+        let change_id_trailers: Vec<&Trailer> = trailers
+            .iter()
+            .filter(|trailer| trailer.key == "Change-Id")
+            .collect();
+
+        // There shouldn't be multiple change-ID fields. So just error out if
+        // there is.
+        if change_id_trailers.len() > 1 {
+            return Err(user_error(format!(
+                "multiple Change-Id footers in commit {}",
+                short_commit_hash(commit_id)
+            )));
+        }
+
+        // The user can choose to explicitly set their own change-ID to
+        // override the default change-ID based on the jj change-ID.
+        if let Some(trailer) = change_id_trailers.first() {
+            // Check the change-id format is correct.
+            if trailer.value.len() != 41 || !trailer.value.starts_with('I') {
+                // Intentionally leave the invalid change IDs as-is.
+                writeln!(
+                    ui.warning_default(),
+                    "warning: invalid Change-Id footer in commit {}",
+                    short_commit_hash(original_commit.id()),
+                )?;
+            }
+
+            // map the old commit to itself
+            old_to_new.insert(original_commit.id().clone(), original_commit.clone());
+            continue;
+        }
+
+        // Gerrit change id is 40 chars, jj change id is 32, so we need padding.
+        // To be consistent with `format_gerrit_change_id_trailer``, we pad with
+        // 6a6a6964 (hex of "jjid").
+        let gerrit_change_id = format!("I6a6a6964{}", original_commit.change_id().hex());
+
+        let new_description = format!(
+            "{}{}Change-Id: {}\n",
+            description.trim(),
+            if trailers.is_empty() { "\n\n" } else { "\n" },
+            gerrit_change_id
+        );
+
+        let new_parents = original_commit
+            .parents()
+            .map(|parent| {
+                let p = parent.unwrap();
```

Comment:
nit: avoid this `.unwrap()` by rewriting the `map()` as a loop or maybe by using `try_collect()`